src/map.rs

```diff
@@ -174,15 +181,23 @@ impl Map {
         // data usage.
         let mut layers = Vec::new();
         let mut properties = HashMap::new();
+
         let mut tilesets = Vec::new();
+        let mut templates = Vec::new();
 
         parse_tag!(parser, "map", {
             "tileset" => |attrs| {
-                let res = Tileset::parse_xml_in_map(parser, attrs, map_path)?;
+                let res = Tileset::parse_xml_in_map(parser, attrs, map_path, &tilesets, &mut templates, None, cache)?;
                 match res.result_type {
                     EmbeddedParseResultType::ExternalReference { tileset_path } => {
-                        let file = File::open(&tileset_path).map_err(|err| TiledError::CouldNotOpenFile{path: tileset_path.clone(), err })?;
-                        let tileset = cache.get_or_try_insert_tileset_with(tileset_path.clone(), || Tileset::new_external(file, Some(&tileset_path)))?;
+                        let tileset = if let Some(ts) = cache.get_tileset(&tileset_path) {
+                            ts
+                        } else {
+                            let file = File::open(&tileset_path).map_err(|err| TiledError::CouldNotOpenFile{path: tileset_path.clone(), err })?;
+                            let tileset = Arc::new(Tileset::new_external(file, Some(&tileset_path), &tilesets, &mut templates, None, cache)?);
+                            cache.insert_tileset(tileset_path.clone(), tileset.clone());
+                            tileset
+                        };
```

Comment:
This change isn't related to templates; I suggest we keep the original `get_or_try` method.

Reply:
Tilesets can contain tiles, which can contain objectgroups which contain objects which can contain templates. So while running inside the closure we need access to the cache (in case we need to process another template's tileset). This would mean that the cache is being mutably borrowed twice, once in the closure, and once as the `&mut self` in cache.

I'm not sure how else we could work around this besides having the closure take the cache as an argument?.